### ISSUE TITLE
another failed uv test

### ISSUE THREAD

**Issue by amiloradovsky** [canonical/raft#87]
Building 0.9.8, one test still fails:
```
uvEnsureDir/statError                                       [ ERROR ]
Error: test/unit/test_uv_os.c:64: assertion failed: uvEnsureDir("/proc/1/root", &errmsg) == UV__ERROR (0 == 1)
Error: child killed by signal 6 (Aborted)
```
My guess, it's because the tests in that group compare for the locale-dependent error message string:
```C
/* If the directory can't be created, an error is returned. */
TEST(uvEnsureDir, mkdirError, NULL, NULL, 0, NULL)
{
    ENSURE_DIR_ERROR("/foobarbazegg", UV__ERROR, "mkdir: permission denied");
    return MUNIT_OK;
}

/* If the directory can't be probed for existence, an error is returned. */
TEST(uvEnsureDir, statError, NULL, NULL, 0, NULL)
{
    ENSURE_DIR_ERROR("/proc/1/root", UV__ERROR, "stat: permission denied");
    return MUNIT_OK;
}

/* If the given path is not a directory, an error is returned. */
TEST(uvEnsureDir, notDir, NULL, NULL, 0, NULL)
{
    ENSURE_DIR_ERROR("/dev/null", UV__ERROR, "not a directory");
    return MUNIT_OK;
}
```

**Comment by freeekanayaka** [comment]
Actually locale has nothing to do this this.

The test fails because it expects an error: ```uvEnsureDir("/proc/1/root", &errmsg) ``` should return ```UV__ERROR``` (where ```UV__ERROR``` is defined as ``1``), but it returns ```0``` instead, hence the ```(0 == 1)``` diagnostic message.

Are you running the tests as root?

**Comment by freeekanayaka** [comment]
Also, what's the output of ```ls -ld /proc/1``` on your system?

**Comment by freeekanayaka** [comment]
And of ```stat /proc/1/root``` too please.

**Comment by amiloradovsky** [comment]
```
$ LANG=C ls -ld /proc/1
dr-xr-xr-x 9 root root 0 Oct 30 14:42 /proc/1/
```
```
$ LANG=C stat /proc/1/root
  File: /proc/1/rootstat: cannot read symbolic link '/proc/1/root': Permission denied

  Size: 0         	Blocks: 0          IO Block: 1024   symbolic link
Device: 5h/5d	Inode: 332292      Links: 1
Access: (0777/lrwxrwxrwx)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2019-10-31 12:51:44.641357192 +0300
Modify: 2019-10-31 12:51:43.049355999 +0300
Change: 2019-10-31 12:51:43.049355999 +0300
 Birth: -
```
```
# LANG=C ls -l /proc/1/root
lrwxrwxrwx 1 root root 0 Oct 31 12:51 /proc/1/root -> /
```

**Comment by freeekanayaka** [comment]
Are you running the tests as root?

**Comment by amiloradovsky** [comment]
No, as dedicated user for the building.

**Comment by freeekanayaka** [comment]
What's the output of:

```
make test/unit/uv && strace ./test/unit/uv --no-fork uvEnsureDir/statError
```

?

**Comment by amiloradovsky** [comment]
[strace.log](https://github.com/canonical/raft/files/3793584/strace.log)

**Comment by freeekanayaka** [comment]
The strace.log you attached contains:

```
write(1, "[ OK    ] [ 0.00005429 / 0.00002"..., 95[ OK    ] [ 0.00005429 / 0.00002694 CPU ]
1 of 1 (100%) tests successful, 0 (0%) test skipped.
```

which means that the test didn't fail. Please attach an strace.log where the test actually fails (e.g. running the command as part of your build/test process).

**Comment by amiloradovsky** [comment]
[guix-build.log](https://github.com/canonical/raft/files/3793893/guix-build.log)
```scheme
    (arguments '(#:configure-flags '("--enable-uv")
                 #:parallel-tests? #f
                 ;; #:tests? #f
                 #:phases
                 (modify-phases
                  %standard-phases
                  (add-before 'check 'strace
                              (lambda _
                                (invoke "make" "test/unit/uv")
                                (invoke "strace" "test/unit/uv"
                                        "--no-fork" "uvEnsureDir/statError"))))))
```

**Comment by amiloradovsky** [comment]
Just inserted
```
(invoke "whoami")
(invoke "stat" "/proc/1/root")
```
and it says
```
nixbld
  File: /proc/1/root -> /
  Size: 0         	Blocks: 0          IO Block: 1024   symbolic link
Device: 2eh/46d	Inode: 817723      Links: 1
Access: (0777/lrwxrwxrwx)  Uid: (  999/  nixbld)   Gid: (30000/  nixbld)
Access: 2019-10-31 13:39:50.778838105 +0000
Modify: 2019-10-31 13:39:50.778838105 +0000
Change: 2019-10-31 13:39:50.778838105 +0000
 Birth: -
```
Apparently the builder has more permissions than my regular unprivileged user.

I'm not sure what the test is supposed to check, but I guess we by now have all the information.

**Comment by freeekanayaka** [comment]
The test is supposed to check that the attempt to create the data dir fail if the process has not enough privileges to do so. I can put it place a workaround which skips the test if /proc/1/root is actually accessible by the test process.

**Comment by amiloradovsky** [comment]
A little tangent:
I've just managed to build `dqlite` on Guix, but only at the cost of disabling the tests in all the packages in the stack, all the way down to `libuv`. Didn't try if it's really working yet. But the need to disable the tests everywhere is kind of worrisome.

**Comment by freeekanayaka** [comment]
The tests all pass on Travis, which is the project's official CI. If have an environment which differs from the one that Travis sets up, please just report your failures like this one and I'll be happy to fix them.

**Comment by freeekanayaka** [comment]
(this is should be fixed in the last v0.9.9 release tag that I just pushed).

**Comment by amiloradovsky** [comment]
OK. Thank you.
